### PR TITLE
Add support for build and publish database project

### PR DESCRIPTION
**Pull request by zijchen**
Introducing 2 new inputs to this action:
- `project-file` - Path to a database project (.sqlproj) file. Project must use the [Microsoft.Build.Sql SDK](https://www.nuget.org/packages/microsoft.build.sql/)
- `build-arguments` - Additional arguments supplied to building the sqlproj with `dotnet build`

The sqlproj will be first built with `dotnet build <project-file> -p:NetCoreBuild=true <build-arguments>`. The output DACPAC file will then be published using existing SqlPackage logic from the DACPAC action.
- This means the workflow runner needs to have both dotnet and SqlPackage installed (only for self-hosted runners, default ones have these pre-installed).
- The current `arguments` input can still be used to specify additional SqlPackage parameters. We may want to rename this in the future to better distinguish between that and `build-arguments`. New features in this PR are intended to be fully backwards-compatible.

Added new tests to cover project build functionality. Also added a step to the `pr-check.yml` to test project build and publish a test sqlproj to Azure.

### CODE REVIEW

File: .github/workflows/pr-check.yml [hunk 1]
```diff
@@ -14,25 +14,43 @@ jobs:
...
-      - uses: azure/login@v1
+      - name: Azure Login
+        uses: azure/login@v1
         with:
           creds: ${{ secrets.AZURE_CREDENTIALS }}
-          
-      - uses: ./
+
+      # Deploy a DACPAC with only a table to server
+      - name: Test DACPAC Action
```
Comment (zijchen) [inline .github/workflows/pr-check.yml:35]: Some of these changes are from #63

File: src/AzureSqlAction.ts [hunk 2]
```diff
@@ -70,6 +94,30 @@ export default class AzureSqlAction {
...
 
+    private async _executeBuildProject(inputs: IBuildAndPublishInputs): Promise<string> {
+        const projectName = path.basename(inputs.projectFile, Constants.sqlprojExtension);
+        const additionalBuildArguments = inputs.buildArguments ?? '';
+        const parsedArgs = DotnetUtils.parseCommandArguments(additionalBuildArguments);
+        let outputDir = '';
+
+        // Set output dir if it is set in the build arguments
+        const outputArg = DotnetUtils.findArgument(parsedArgs, "--output", "-o");
+        if (!outputArg) {
```
Comment (kisantia) [inline src/AzureSqlAction.ts:105]: shouldn't this be `if (outputArg)` since it should get set if it isn't undefined?

**Review by kisantia — APPROVED** [review]

File: src/AzureSqlAction.ts (hunk 2 above)
Comment (zijchen) [reply src/AzureSqlAction.ts:105]: No we're actually setting outputDir to this value if it was passed in. Renamed the variable to be more clear.

File: __tests__/AzureSqlAction.test.ts [hunk 3]
```diff
@@ -95,5 +149,14 @@ function getInputs(actionType: ActionType) {
                 additionalArguments: '-t 20'
             } as ISqlActionInputs;
         }
+        case ActionType.BuildAndPublish: {
+            return {
+                serverName: 'testServer.database.windows.net',
+                actionType: ActionType.BuildAndPublish,
+                connectionString: new SqlConnectionStringBuilder('Server=testServer.database.windows.net;Initial Catalog=testDB;User Id=testUser;Password=testPassword'),
```
Comment (udeeshagautam) [inline __tests__/AzureSqlAction.test.ts:156]: Can you please change these to word "placeholder" too

File: action.yml [hunk 4]
```diff
@@ -1,20 +1,26 @@
...
-    description: 'Name of the Azure SQL Server name, like Fabrikam.database.windows.net.'
+    description: 'Name of the Azure SQL Server name, like Fabrikam.database.windows.net.'
     required: false
   connection-string:
-    description:  'The connection string, including authentication information, for the Azure SQL Server database.'
+    description:  'The connection string, including authentication information, for the Azure SQL Server database.'
     required: true
   dacpac-package:
-    description: 'Path to DACPAC file to deploy'
+    description: 'Path to DACPAC file to deploy'
```
Comment (Benjin) [inline action.yml:11]: Whitespace diff.  Huh.

File: src/DotnetUtils.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+     * @param args A dictionary where keys are arg names and values are arg values.
+     * @param argumentToFind The argument to look for.
+     * @param alias The alias of the argument to look for.
+     */
+    public static async findArgument(args: object, argumentToFind: string, alias?: string): Promise<string | undefined> {
+        if (alias === undefined) {
+            return args[argumentToFind];
+        } else {
+            return args[argumentToFind] ?? args[alias];
+        }
```
Comment (Benjin) [inline src/DotnetUtils.ts:52]: Maybe worth cleaning up as `return alias ? args[alias] : args[argumentToFind]`

or

```
argumentToFind = alias ?? argumentToFind;
return args[argumentToFind];
```

File: __tests__/DotnetUtils.test.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+        });
+    });
+
+    it ('should find arguments parsed by parseCommandArguments', async () => {
+        const dotnetParams = `-o C:\\test\\ --configuration 'Debug' --verbose --test "test value"`;
+        const parsedArgs = await DotnetUtils.parseCommandArguments(dotnetParams);
+
+        expect(await DotnetUtils.findArgument(parsedArgs, '')).toEqual(undefined);
+        expect(await DotnetUtils.findArgument(parsedArgs, '-o')).toEqual('C:\\test\\');
+        expect(await DotnetUtils.findArgument(parsedArgs, '--output', '-o')).toEqual('C:\\test\\');
```
Comment (Benjin) [inline __tests__/DotnetUtils.test.ts:25]: Are these checks redundant due to `-o`?

**Review by Benjin — APPROVED** [review]

File: action.yml (hunk 4 above)
Comment (zijchen) [reply action.yml:11]: In VS Code the original "spaces" were actually some invalid characters

**Review by udeeshagautam — APPROVED** [review]

File: src/DotnetUtils.ts (hunk 5 above)
Comment (zijchen) [reply src/DotnetUtils.ts]: Updated. Logic is slightly different than either of your suggestions as we want to look for `args[argumentToFind]` first regardless if alias is set.

File: __tests__/AzureSqlAction.test.ts (hunk 3 above)
Comment (zijchen) [reply __tests__/AzureSqlAction.test.ts]: Updated

**Pull request merged**